wp-content/plugins/tapestry/controller/class.tapestry-controller.php

```diff
@@ -3,4 +3,106 @@
  * Add/update/retrieve Tapestry post and its child nodes
  * 
  */
-class TapestryController {}
\ No newline at end of file
+
+class TapestryController {
+    const POST_TYPES = [
+        'TAPESTRY' => 'tapestry',
+        'TAPESTRY_NODE' => 'tapestry_node'
+    ];
+    const ERRORS = [
+        'INVALID_POST_ID' => [
+            'MESSAGE' => 'PostID is invalid',
+            'STATUS' => ['status' => 404]
+        ]
+    ];
+
+    /**
+     * Update Tapestry nodes first then
+     * Update the existing Tapestry if the postId is provided
+     * Otherwise, a new Tapestry will be created
+     * 
+     * @param  Object @tapestry The Tapestry data
+     * @param  Number @postId The postId of the Tapestry
+     * @return Object @tapestry
+     */
+    public function updateTapestry($tapestry, $postId = null) {
+        // TODO: use isValidPostID() utlility function
+        if (!isset($postId)) {
+            $postId = $this->updateWordPressPost($tapestry, 'tapestry');
+        }
+        $this->updateNodes($tapestry->nodes, $postId);
+
+        if (!isset($tapestry->rootId)) {
+            $tapestry->rootId = $tapestry->nodes[0]->id;
+        }
+        $tapestry->nodes = $this->getNodeIds($tapestry->nodes);
+
+        update_post_meta($postId, 'tapestry', $tapestry);
+        return $tapestry;
+    }
+
+    private function updateNodes($nodes, $postId) {
+        foreach ($nodes as $node) {
+            if (!isset($node->id)) {
+                $nodePostId = $this->updateWordPressPost($node, 'tapestry_node');
+                $metadata = $this->makeMetadata($node, $nodePostId);
+                $node->id = add_post_meta($postId, 'tapestry_node', $metadata);
+            } else {
+                $metadata = get_metadata_by_mid('post', $node->id)->meta_value;
+                $nodePostId = $metadata->post_id;
+            }
+            update_post_meta($nodePostId, 'tapestry_node_data', $node);
+        }
+    }
+
+    private function updateWordPressPost($post, $type, $postId = null) {
```

Comment:
Please avoid using phrases that is the name of the application or plugin because it just makes the name longer. For example, this function should not have `WordPress` as part of it.